src/librustc/mir/interpret/value.rs

```diff
@@ -134,6 +132,12 @@ impl<Tag> fmt::Display for Scalar<Tag> {
 }
 
 impl<'tcx> Scalar<()> {
+    #[inline]
+    fn check_data(data: u128, size: u8) {
+        assert_eq!(truncate(data, Size::from_bytes(size as u64)), data,
```

Comment:
If perf still acts up, this can also be a debug_assert. Users can't act on it anyway and if we get weird crashes, in a dev build we'll immediately see it.

Reply:
I'd feel better if we could have assertions in release builds at least for either insertion or extraction... but I agree.